### PR TITLE
Add test of distributed HDFS connector with hard bounces and validate exactly once delivery.

### PR DESCRIPTION
**Pull request by ewencp**

### CODE REVIEW

**Comment by ewencp** [comment]
@Ishiihara or @gwenshap This now passes consistently for me with confluentinc/kafka-connect-hdfs#22. Includes some refactoring in `connect_hdfs_test.py` to make a bunch of stuff reusable (and also reduces some effectively duplicated code from the existing tests).

File: muckrake/services/templates/connect_log4j.properties [hunk 1]
```diff
@@ -16,7 +16,7 @@
 ##
 
 # Define the root logger with appender file
-log4j.rootLogger = {{ log_level|default("INFO") }}, FILE
```
Comment (guozhangwang) [inline muckrake/services/templates/connect_log4j.properties]: Is this change intentional?

File: muckrake/tests/connect_hdfs_distributed_test.py [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+                with node.account.monitor_log(self.connect_sink.LOG_FILE) as monitor:
+                    self.connect_sink.start_node(node)
+                    monitor.wait_until("Starting connectors and tasks using config offset", timeout_sec=90,
+                                       err_msg="Kafka Connect worker didn't successfully join group and start work")
+                self.logger.info("Bounced Kafka Connect on %s and rejoined in %f seconds", node.account, time.time() - started)
+                # Give additional time for recovery. We need this for two reasons. First, the consumer groups need time
+                # to recover as they will have to wait for a session timeout. Second, the leases for HDFS need to expire
+                # in case they were held during the hard kill. The latter requires a longer timeout (defaults to 1 min)
+                # and we need to give it some additional time to actually make progress in writing data.
+                time.sleep(75)
```
Comment (guozhangwang) [inline muckrake/tests/connect_hdfs_distributed_test.py]: Just curious about the magic number here, is it related to the lease timeout?

File: muckrake/services/templates/connect_log4j.properties (hunk 1 above)
Comment (ewencp) [inline muckrake/services/templates/connect_log4j.properties]: Yes. Putting the root logger at levels more detailed than INFO made the logs really difficult to use for debugging because there was too much noise from other stuff (e.g. lots of HDFS logging). Instead, specifically targetting Kafka Connect stuff (the additions in this file) is much more useful.

File: muckrake/tests/templates/hdfs-sink.properties [hunk 3]
```diff
@@ -1,11 +1,11 @@
...
-tasks.max=1
+tasks.max={{ tasks }}
 topics={{ topic }}
-hdfs.url=hdfs://{{ namenode }}:9000
+hdfs.url=hdfs://{{ hadoop.master_host }}:9000
 hadoop.conf.dir=/mnt
 hadoop.home=/opt/hadoop-cdh
 flush.size={{ flush_size }}
-format= {{ format }}
+format={{format}}
```
Comment (guozhangwang) [inline muckrake/tests/templates/hdfs-sink.properties]: Is this intentional?

File: muckrake/tests/connect_hdfs_distributed_test.py (hunk 2 above)
Comment (ewencp) [inline muckrake/tests/connect_hdfs_distributed_test.py]: Yes, the lease timeout is the key factor -- 75 is a healthy buffer above the 1 min timeout. In theory we could make this test faster if we reduced the lease time.

File: muckrake/tests/templates/hdfs-sink.properties (hunk 3 above)
Comment (ewencp) [inline muckrake/tests/templates/hdfs-sink.properties]: Not specifically, though it has no impact. However, avoiding extra spaces in these files is generally a good idea since connectors may not always trim the string before trying to use it.

**Comment by guozhangwang** [comment]
LGTM.
